kratos/containers/system_vector.h

```diff
@@ -0,0 +1,352 @@
+#include "containers/sparse_contiguous_row_graph.h"
+
+namespace Kratos
+{
+///@addtogroup ApplicationNameApplication
+///@{
+
+///@name Kratos Globals
+///@{
+
+///@}
+///@name Type Definitions
+///@{
+
+///@}
+///@name  Enum's
+///@{
+
+///@}
+///@name  Functions
+///@{
+
+///@}
+///@name Kratos Classes
+///@{
+
+/// Provides a SystemVector which implements FEM assemble capabilities
+template<class TDataType=double, class TIndexType=std::size_t>
+class SystemVector
+{
+public:
+    ///@name Type Definitions
+    ///@{
+    typedef TIndexType IndexType;
+
+    /// Pointer definition of SystemVector
+    KRATOS_CLASS_POINTER_DEFINITION(SystemVector);
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    SystemVector(const SparseGraph<IndexType>& rGraph){
+        mData.resize(rGraph.Size(),false);
+    }
+
+    SystemVector(const SparseContiguousRowGraph<IndexType>& rGraph){
+        mData.resize(rGraph.Size(),false);
+    }
+
+    SystemVector(IndexType size){
+        mData.resize(size,false);
+    }
+
+    /// Copy constructor.
+    explicit SystemVector(const SystemVector<TDataType,TIndexType>& rOtherVector){
+        mData.resize(rOtherVector.size(),false);
+
+        IndexPartition<IndexType>(size()).for_each([&](IndexType i){
+            (*this)[i] = rOtherVector[i];
+        });
+    }
+
+    /// Destructor.
+    virtual ~SystemVector(){}
+
+    ///@}
+    ///@name Operators
+    ///@{
+    void Clear()
+    {
+        mData.clear();
+    }
+
+    void SetValue(const TDataType value)
+    {
+        IndexPartition<IndexType>(mData.size()).for_each([&](IndexType i){
+            mData[i] = value;
+        });
+    }
+
+    IndexType size() const
+    {
+        return mData.size();
+    }
+
+    TDataType& operator()(IndexType I){
+        return mData[I];
+    }
+
+    const TDataType& operator()(IndexType I) const{
+        return mData[I];
+    }
+
+    TDataType& operator[](IndexType I){
+        return mData[I];
+    }
+
+    const TDataType& operator[](IndexType I) const{
+        return mData[I];
+    }
+
+    void Add(const double factor,
+             const SystemVector& rOtherVector
+            )
+    {
+        IndexPartition<IndexType>(size()).for_each([&](IndexType i){
+            (*this)[i] += factor*rOtherVector[i];
+        });
+    }
+
+    /// Assignment operator.
+    SystemVector& operator=(SystemVector const& rOtherVector){
+        IndexPartition<IndexType>(size()).for_each([&](IndexType i){
+            (*this)[i] = rOtherVector[i];
+        });
+    }
+
+
+    void operator+=(const SystemVector& rOtherVector)
+    {
+        IndexPartition<IndexType>(size()).for_each([&](IndexType i){
+            (*this)[i] += rOtherVector[i];
+        });
+    }
+
+    void operator-=(const SystemVector& rOtherVector)
+    {
+        IndexPartition<IndexType>(size()).for_each([&](IndexType i){
+            (*this)[i] -= rOtherVector[i];
+        });
+    }
+
+    void operator*=(const TDataType& multiplier_factor)
```

Comment:
I'd remove the reference to the `TDataType` in here.

Reply:
ok